libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp

```diff
@@ -0,0 +1,121 @@
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include "AP_BattMonitor.h"
+#include "AP_BattMonitor_FuelFlow.h"
+#include <GCS_MAVLink/GCS.h>
+#include <stdio.h>
+
+/*
+  "battery" monitor for liquid fuel flow systems that give a pulse on
+  a pin for fixed volumes of fuel.
+
+  this driver assumes that BATTx_AMP_PERVLT is set to give the
+  number of millilitres per pulse.
+
+  Output is:
+
+    - current in Amps maps to in litres/hour
+    - consumed mAh is in consumed millilitres
+    - fixed 1.0v voltage
+ */
+extern const AP_HAL::HAL& hal;
+
+/// Constructor
+AP_BattMonitor_FuelFlow::AP_BattMonitor_FuelFlow(AP_BattMonitor &mon,
+                                                 AP_BattMonitor::BattMonitor_State &mon_state,
+                                                 AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+}
+
+/*
+  handle interrupt on an instance
+ */
+void AP_BattMonitor_FuelFlow::irq_handler(uint8_t pin, bool pin_state, uint32_t timestamp)
+{
+    if (irq_state.last_pulse_us == 0) {
+        irq_state.last_pulse_us = timestamp;
+        return;
+    }
+    irq_state.pulse_count++;
+    irq_state.total_us += (timestamp - irq_state.last_pulse_us);
+    irq_state.last_pulse_us = timestamp;
+}
+
+/*
+  read - read the "voltage" and "current"
+*/
+void AP_BattMonitor_FuelFlow::read()
+{
+    int8_t pin = _params._curr_pin;
+    if (last_pin != pin) {
+        // detach from last pin
+        if (last_pin != -1) {
+            hal.gpio->detach_interrupt(last_pin);
+        }
+        // attach to new pin
+        last_pin = pin;
+        if (last_pin > 0) {
+            hal.gpio->pinMode(last_pin, HAL_GPIO_INPUT);
+            if (!hal.gpio->attach_interrupt(
+                    last_pin,
+                    FUNCTOR_BIND_MEMBER(&AP_BattMonitor_FuelFlow::irq_handler, void, uint8_t, bool, uint32_t),
+                    AP_HAL::GPIO::INTERRUPT_RISING)) {
+                gcs().send_text(MAV_SEVERITY_WARNING, "FuelFlow: Failed to attach to pin %u", last_pin);
+            }
+        }
+    }
+
+    uint32_t now_us = AP_HAL::micros();
+    if (_state.last_time_micros == 0) {
+        // need initial time, so we can work out expected pulse rate
+        _state.last_time_micros = now_us;
+        return;
+    }
+    float dt = (now_us - _state.last_time_micros) * 1.0e-6;
+
+    if (dt < 1 && irq_state.pulse_count == 0) {
```

Comment:
if dt > 1 amd we haven't gotten healthy is this an indicator that we should be setting healthy to false?

Reply:
no, as the comment states, we don't want to be unhealthy when in pre-arm, with zero throttle

Reply:
That makes sense. Do we want to just set the `_state.healthy` in the constructor (or init) then? There isn't any value in doing it each loop for the driver.

Reply:
done